File: src/com/leetcode/Medium/ReorderList.java
package com.leetcode.Medium;

// 143. Reorder List
public class ReorderList {
    static class ListNode {
        int val;
        ListNode next;

        ListNode(int x) {
            val = x;
        }
    }

    private ListNode reverse(ListNode slow) {
        ListNode prev = null;
        while (slow != null) {
            ListNode node = slow.next;
            slow.next = prev;
            prev = slow;
            slow = node;
        }
        return prev;
    }

    public void reorderList(ListNode head) {
        if (head == null)
            return;

        ListNode fast = head, slow = head;
        while (fast != null && fast.next != null) {
            slow = slow.next;
            fast = fast.next.next;
        }

        ListNode headOfFirst = head;
        ListNode headOfSecond = reverse(slow);

        while (headOfFirst != null && headOfSecond != null) {
            ListNode temp = headOfFirst.next;
            headOfFirst.next = headOfSecond;
            headOfFirst = temp;

            temp = headOfSecond.next;
            headOfSecond.next = headOfFirst;
            headOfSecond = temp;
        }

        if (headOfFirst != null)
            headOfFirst.next = null;
    }
}

File: JavaScript/Medium/ReorderList.js
/**
 * 143. Reorder List
 * @param {ListNode} head
 * @return {void} Do not return anything, modify head in-place instead.
 */
const reorderList = function (head) {
  if (head == null)
    return;

  let fast = head, slow = head;
  while (fast != null && fast.next != null) {
    slow = slow.next;
    fast = fast.next.next;
  }

  let headOfFirst = head;
  let headOfSecond = reverse(slow);

  while (headOfFirst != null && headOfSecond != null) {
    let temp = headOfFirst.next;
    headOfFirst.next = headOfSecond;
    headOfFirst = temp;

    temp = headOfSecond.next;
    headOfSecond.next = headOfFirst;
    headOfSecond = temp;
  }

  if (headOfFirst != null)
    headOfFirst.next = null;
};

const reverse = function (node) {
  let prev = null;
  while (node != null) {
    let next = node.next;
    node.next = prev;
    prev = node;
    node = next;
  }
  return prev;
};